iavl/iavl_tree.go

```diff
@@ -41,139 +135,303 @@ func NewIAVLTree(cacheSize int, db dbm.DB) *IAVLTree {
+func (t *IAVLTree) Version() int {
+	//fmt.Printf("Version ")
+	root := t.GetRoot(t.version)
+	if root == nil {
+		return 0
+	}
+	return root.version
 }
 
 func (t *IAVLTree) Has(key []byte) bool {
-	if t.root == nil {
+	//fmt.Printf("Has ")
+	root := t.GetRoot(t.version)
+	if root == nil {
 		return false
 	}
-	return t.root.has(t, key)
+	return root.has(t, key)
 }
 
+// Proof of the latest key
 func (t *IAVLTree) Proof(key []byte) (value []byte, proofBytes []byte, exists bool) {
-	value, proof := t.ConstructProof(key)
+	//fmt.Printf("Proof ")
+	value, _, proof := t.ConstructProof(key, t.version)
 	if proof == nil {
+		//fmt.Printf("Missing Proof\n")
 		return nil, nil, false
 	}
 	proofBytes = wire.BinaryBytes(proof)
 	return value, proofBytes, true
 }
 
+// Proof of a key at a specific version
+func (t *IAVLTree) ProofVersion(key []byte, version int) (value []byte, proofBytes []byte, exists bool) {
+	//fmt.Printf("ProofVersion\n")
+	value, _, proof := t.ConstructProof(key, version)
+	if proof == nil {
+		return nil, nil, false
+	}
+	proofBytes = wire.BinaryBytes(proof)
+	return value, proofBytes, true
+}
+
+// Set a value with a new key or an existing one
 func (t *IAVLTree) Set(key []byte, value []byte) (updated bool) {
-	if t.root == nil {
-		t.root = NewIAVLNode(key, value)
+	//fmt.Printf("Set %X/%X ", key, value)
+
+	root := t.GetRoot(t.version)
+	if root == nil {
+		//fmt.Printf("Initializing Tree\n")
+
+		SetValue(t.roots, 0, NewIAVLNode(key, value, t.version))
+
 		return false
 	}
-	t.root, updated = t.root.set(t, key, value)
+
+	//fmt.Printf("Added new Root\n")
+	root.validate(t)
+	root, updated = root.set(t, key, value)
+	root.validate(t)
+
+	SetValue(t.roots, 0, root)
+
+	root = t.GetRoot(t.version) // TEST
+
 	return updated
 }
 
+// Hash returns the root hash for the tree
+// Should not be exposed, since it is just a step in saving
 func (t *IAVLTree) Hash() []byte {
-	if t.root == nil {
+	//fmt.Printf("Hash ")
+	root := t.GetRoot(t.version)
+	if root == nil {
 		return nil
 	}
-	hash, _ := t.root.hashWithCount(t)
+	hash, _ := root.hashWithCount(t)
 	return hash
 }
 
+// HashWithCount returns the root hash for the tree and the size of the tree
+// Should not be exposed, since it is just a step in saving
 func (t *IAVLTree) HashWithCount() ([]byte, int) {
-	if t.root == nil {
+	//fmt.Printf("HashWithCount ")
+	root := t.GetRoot(t.version)
+	if root == nil {
 		return nil, 0
 	}
-	return t.root.hashWithCount(t)
+	return root.hashWithCount(t)
 }
 
+// Save this version of the tree
 func (t *IAVLTree) Save() []byte {
-	if t.root == nil {
+
+	//t.ndb.mtx.Lock()
+	//defer t.ndb.mtx.Unlock()
+
+	//fmt.Printf("****** Save\n")
+	//t.PrintRoots()
+
+	// Might be the same
+	first := t.roots.Front()
+	if first == nil {
+		// Empty list
+		return nil
+	}
+
+	firstNode := first.Value.(*IAVLNode)
+
+	if firstNode == nil {
+		// Null root
 		return nil
+
+	} else if t.ndb != nil {
+		firstNode.save(t)
```

Comment:
It feels weird to me that the node takes the tree as an arg. I know this is how it was, but I wonder if we can remove that. eg. writePersistBytes takes the tree but doesn't seem to use it.